### PR TITLE
feat(traffic_light_multi_camera_fusion): take confidence into consideration

### PR DESCRIPTION
**Pull request by tzhong518**
## Description

The purpose of this PR is to avoid mistrial resulted from low-confidence misdetection.
Before, the multi camera fusion is based only on the size and truncation state of the rois. As a consequence, if the misdetected roi is not truncated and has the biggest area among the group, the fusion result could be misled to green as below.
image
Trusting the results with the highest confidence instead of area could avoid such problem. And it also seems reliable in other situations as far as we observed.
image
left: detection result
right: fusion result

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [ ] I've confirmed the [contribution guidelines].
- [ ] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

**Review by yuki-takagi-66 — COMMENTED** [review]
Thank you for the PR.
I've commented on a minor matter, but as abstracted, it is LGTM.

File: perception/traffic_light_multi_camera_fusion/src/node.cpp [hunk 1]
```diff
@@ -82,13 +82,17 @@ int compareRecord(const traffic_light::FusionRecord & r1, const traffic_light::F
...
+    for_each(r1.signal.elements.begin(), r1.signal.elements.end(), [&sum_1](const auto & elem) {
+      sum_1 = sum_1 + elem.confidence;
+    });
+    double sum_2 = 0;
+    for_each(r2.signal.elements.begin(), r2.signal.elements.end(), [&sum_2](const auto & elem) {
+      sum_2 = sum_2 + elem.confidence;
+    });
+    double confidence_1 = sum_1 / r1.signal.elements.size();
+    double confidence_2 = sum_2 / r2.signal.elements.size();
+    return confidence_1 < confidence_2 ? -1 : 1;
```
[the remark this answers is not in the document]
Comment (tzhong518) [reply perception/traffic_light_multi_camera_fusion/src/node.cpp:87]: One signal can be composed of multiple elements(bulbs), e.g. in
Screenshot from 2024-02-16 13-39-00
size() will return 3.
But thanks to your comment, I realized the elements of the same signal share the same confidence, so it is sufficient to compare the confidence of the first element: https://github.com/autowarefoundation/autoware.universe/pull/6267/commits/564bf45b9b1a500d652638ba3ed4f290b4e4d613

[the remark this answers is not in the document]
Comment (yuki-takagi-66) [reply perception/traffic_light_multi_camera_fusion/src/node.cpp:87]: Thank you for the kind response.
I had miss understood.

**Review by yuki-takagi-66 — APPROVED** [review]
LGTM

**Review by miursh — APPROVED** [review]

**Pull request merged**